### PR TITLE
SLVUU-117 add type guard for layoutJSON

### LINKED ISSUE #117: Empty JSON being saved to local storage
### Background

Developers at UBS are sometimes experiencing empty JSON is being saved to local storage. It's not actively causing problems, but it could be indicative of problems earlier in the chain (e.g. React -> JSON serialization). We should ensure an empty layout definition does not get saved.

### Work Required
- Put (a) guard clause(s) in place to stop saving empty or undefined layouts for both applicationLayout and layout
- Trigger error notificationb when save layout action is executed with empty layout {}

### Notes

Related: https://github.com/ScottLogic/finos-vuu/pull/115#discussion_r1400319352

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
### Changes: 
- add type guard for LayoutJSON 
- check layout is LayoutJSON before saving and show console error and error notification if it is not
- check application layout is LayoutJSON before saving and show console error if it is not

### CODE REVIEW

**Review by cfisher-scottlogic — COMMENTED** [review]
Do we have any automated tests for this? How much do we gain by making one at this stage?

File: vuu-ui/packages/vuu-layout/src/utils/typeOf.ts [hunk 1]
```diff
@@ -23,3 +23,5 @@ export function typeOf(element?: LayoutModel | WithType): string | undefined {
 }
 
 export const isTypeOf = (element: ReactElement, type: string) => typeOf(element) === type;
+
+export const isLayoutJSON = (layout: LayoutJSON): layout is LayoutJSON=> "type" in layout;
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/utils/typeOf.ts:27]: Unfortunately that won't work in our case as typeof can only return '"string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"' so in our case it typeof layout will return "object"

**Comment by vferraro-scottlogic** [comment]
> Do we have any automated tests for this? How much do we gain by making one at this stage?

I am going to add some unit tests for useLayoutManager as there is none at the moment.
e2e tests-wise we have [this ticket](https://github.com/ScottLogic/finos-vuu/issues/50) that will take care of that

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    expect(persistence.loadApplicationJSON).toHaveBeenCalled();
+    expect(result.current.applicationJson).toBe(initialApplicationJSON);
+  });
+
+  describe("saveApplicationLayout", () => {
+    it("calls saveApplicationJSON when layout is valid", () => {
+      const { result } = renderHook(useLayoutManager, { wrapper });
+
+      vi.mocked(persistence.saveApplicationJSON).mockResolvedValueOnce();
+      vi.mocked(isLayoutJSON).mockReturnValue(true);
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:127]: Do we have to mock the result of `isLayoutJSON`? I think it would be best to test that the function is working as expected, so would that mean writing tests for the utils file?

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    it("calls createLayout when layout is valid and path is resolved", () => {
+      const { result } = renderHook(() => useLayoutManager(), { wrapper });
+
+      vi.mocked(persistence.createLayout).mockResolvedValueOnce(metadata);
+      vi.mocked(isLayoutJSON).mockReturnValue(true);
+      vi.mocked(resolveJSONPath).mockReturnValue(newLayout);
+
+      result.current.saveLayout(metadata);
+
+      expect(persistence.createLayout).toHaveBeenCalled();
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:163]: Is it worth using `toHaveBeenCalledWith` here, like you did above?

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    });
+
+    it("doesn't call createLayout, triggers error notification and logs error when layout path can't be resolved ", () => {
+      const { result } = renderHook(() => useLayoutManager(), { wrapper });
+      const { notify } = useNotifications();
+
+      vi.mocked(persistence.createLayout).mockResolvedValueOnce(metadata);
+      vi.mocked(resolveJSONPath).mockReturnValue(undefined);
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:171]: Might be good to mock a positive result for `isLayoutJSON` here, just to ensure that has no bearing on the outcome of this test.

**Review by pling-scottlogic — COMMENTED** [review]
A few comments on the tests. I haven't looked too closely at the setup/mocking because it's not something I'm familiar with - I'm happy with that side of it if you're satisfied it's all working properly.

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 4 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:171]: good point

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 2 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:127]: I think it makes sense to mock it here as the test should only be concerned with testing the functionality of the hook. I would test that separately together with the other functionalities from utils file

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 3 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:163]: definitely worth using toHaveBeenCalledWith here too

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 2 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:129]: I have added a couple of test cases for it in a separate test file

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 3 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx]: done

File: vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx (hunk 4 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/test/layout-management/useLayoutManager.test.tsx:176]: done

File: vuu-ui/packages/vuu-layout/test/typeOf.test.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { describe, expect, it } from "vitest";
+import { LayoutJSON, isLayoutJSON } from "../src";
+
+const validLayout: LayoutJSON = {
+  type: "Stack",
+};
+
+const invalidLayout: unknown = {
+  name: "invalid-layout-test",
+};
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-layout/test/typeOf.test.ts:10]: What would happen if we had an object with all the required fields of a `LayoutJSON`, but isn't declared as such? Something like this:

```
const mayOrMayNotBeValidLayout: unknown = {
  type: "Stack",
};
```

Or if there's no type declaration at all?

**Review by pling-scottlogic — COMMENTED** [review]
I feel like there should be some more edge cases for the `isLayoutJSON` tests, but I can't think of any.

File: vuu-ui/packages/vuu-layout/test/typeOf.test.ts (hunk 5 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/typeOf.test.ts:10]: You will get a compile-time error if you give it an object whose type is declared as unknown ("Argument of type 'unknown' is not assignable to parameter of type 'LayoutJSON'") unless it gets cast as LayoutJSON. 
At run-time any object that has a "type" property will pass the validation

Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/typeOf.test.ts:10]: object with no type won't give you a compile-time error as long as there is sufficient overlap between that object inferred type and LayoutJSON (meaning that that object has a "type" property)

**Review by pling-scottlogic — APPROVED** [review]

**Comment by pling-scottlogic** [comment]
[PR merged into Finos](https://github.com/finos/vuu/pull/1124)
